### PR TITLE
Remove protocol restriction for SRV records in cloudflare_dns module. Fixes #36708

### LINKED ISSUE #36708: Cloudflare DNS does not allow for tls protocol type
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
cloudflare_dns

##### ANSIBLE VERSION
```
ansible 2.4.3.0
  python version = 2.7.14 (default, Feb 17 2018, 10:42:17) [GCC 7.3.1 20180130 (Red Hat 7.3.1-2)]
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
N/A

##### SUMMARY
Cloudflare DNS module does not allow for SRV-records of type tls - only tcp and udp are allowed. Cloudflare also supports tls-type.

##### STEPS TO REPRODUCE

```yaml
- cloudflare_dns:
    domain: my.com
    service: foo
    proto: tls
    port: 3500
    priority: 10
    weight: 20
    type: SRV
    value: fooserver.my.com
```

##### EXPECTED RESULTS
Not failing

##### ACTUAL RESULTS
```
TASK [dns : cloudflare_dns] ***********************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "value of proto must be one of: tcp,udp, got: tls"}
```

### PR DESCRIPTION
**Pull request by DennisGlindhart**
##### SUMMARY
Remove protocol restriction for SRV records in cloudflare_dns module. Fixes #36708

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
cloudflare_dns

##### ANSIBLE VERSION
```
ansible 2.4.3.0
  python version = 2.7.14 (default, Feb 17 2018, 10:42:17) [GCC 7.3.1 20180130 (Red Hat 7.3.1-2)]

```

### CODE REVIEW

File: lib/ansible/modules/net_tools/cloudflare_dns.py [hunk 1]
```diff
@@ -44,7 +44,7 @@
   proto:
     description: Service protocol. Required for C(type=SRV)
     required: false
-    choices: [ 'tcp', 'udp' ]
+    choices: [ 'tcp', 'udp', 'tls' ]
```
Comment (Akasurde) [inline lib/ansible/modules/net_tools/cloudflare_dns.py]: Could you please mention like this in description 

```
option 'tls' is added version 2.6
```

**Comment by andreaso** [comment]
Ok, up until now I had simply assumed that _tcp_ and _udp_ were the only two allowed values for the SRV record protocol field. Yet, taking a closer look at [RFC 2782](https://tools.ietf.org/html/rfc2782) those two protocols are simply listed as _"at present the most useful values for this field"_.

Likewise CloudFlare appear to accept all kinds of values in the protocol position, not simply _tcp_, _udp_, and _tls_.

@DennisGlindhart Is there any specific and/or common use case for naming the _tls_ protocol in SRV records? That is, does it make the most sense to explicit add _tls_ as an allowed parameter value, or does it make more sense to fully remove the present restrictions?

**Comment by DennisGlindhart** [comment]
@andreaso According to RFC 2782 all kinds of values are allowed, yes. Although, using the Cloudflare Admin-UI in the browser, it seems that only tcp, udp and tls are possible to create in CF. It might be worth investigating whether or not CF allows for other types when using the API instead.

If CF really does support arbitrary values, I think it would make sense to remove the restriction completely.

I'll try to investigate further and adjust PR accordingly.

**Comment by andreaso** [comment]
Using the API (or more specifically, by using a locally modified cloudflare_dns module) I managed to create a SRV record with the made-up _zebra_ protocol.

On the other hand, there is a benefit in keeping a restrictive parameter list, as it makes it easier to use the module properly. Especially since it's to my knowledge is very unusual for people to need SRV records for arbitrary protocols.

Hence back to my original question, what's the use case for the _tls_ protocol in this context?

**Comment by DennisGlindhart** [comment]
@andreaso Yep - I was also able to create some "random" protocol via the API.
My particular case for tls was setting up some Office 365/Lync
``
SRV _sip._tls.example.com > sipdir.online.lync.com
``
(Example here: https://support.microsoft.com/en-us/help/2566790/troubleshooting-skype-for-business-online-dns-configuration-issues-in)

So TL;DR: Cloudflare API allows us to follow the RFC and use arbitrary protocol. Although tcp, udp and tls are probably enough for 99% of users, there are valid scenarios for other protocols as well. (Maybe we will soon see more protocols used (i.e SCTP or DCCP) when NAT is more or less going away with IPv6 currently preventing use of other protocols than TCP and UDP)

I don't think removing the restriction makes the module that much harder to use, and following the standard  is more important when we have the possibility. We can write the "normal" values (tcp & udp) in the description. I think people using this module will mostly be copying the values from some reference (like the link above) - or alternatively - know what they are doing.

I'm thinking of just remove the restrictions then - Agree?

**Comment by andreaso** [comment]
Thanks for helping me sort out what is what here! At this point I can see the merit with both approaches, and I'll be willing to approve whichever one you choose to move forward with in this pull request.

**Comment by DennisGlindhart** [comment]
I have now removed the restriction altogether so we follow the RFC.

I've also "rebased" / removed my previous commits (adding tls) so this is one clean commit and changed the title of the PR to be more correct. I think this should be ready for merge now.

Thanks for all your help!

cc @andreaso

File: lib/ansible/modules/net_tools/cloudflare_dns.py [hunk 2]
```diff
@@ -42,9 +42,8 @@
     required: false
     default: "1"
   proto:
-    description: Service protocol. Required for C(type=SRV)
+    description: Service protocol. Required for C(type=SRV). Common values are tcp and udp. (Before Ansible 2.6 only tcp and udp were available)
```
Comment (Akasurde) [inline lib/ansible/modules/net_tools/cloudflare_dns.py]: Missing . at end.

Comment (Akasurde) [inline lib/ansible/modules/net_tools/cloudflare_dns.py]: Also, add this note is [CHANGELOG.md](https://github.com/ansible/ansible/blob/devel/CHANGELOG.md#minor-changes) telling end-user about this.

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by andreaso** [comment]
shipit

**Comment by andreaso** [comment]
@Akasurde: Happy as well?

**Comment by DennisGlindhart** [comment]
@Akasurde I adjusted according to your review/change requests and pushed a new commit replacing the existing one with the intention to to keep history simple, but the automated bot/checks does not seem to have picked up the change (maybe because it wasn't a new commit on top of the existing one).
TL;DR Your requested changes have been implemented, so this should AFAIK be ready for merge :)

**Review by Akasurde — APPROVED** [review]

**Review by gundalow — APPROVED** [review]
Assuming `udp` and `UdP` work the same. Otherwise use tolower

**Comment by andreaso** [comment]
As a general rule DNS names are compared without considering case.

When it comes to SRV records [RFC 2782](https://tools.ietf.org/html/rfc2782) explicitly  states that _"The Proto is case insensitive."_

**Comment by gundalow** [comment]
Thanks for confirming

**Comment by Akasurde** [comment]
rebuild_merge

**Comment by Akasurde** [comment]
@DennisGlindhart @andreaso @gundalow Thanks for contribution.
